### PR TITLE
[SPARK-41945][CONNECT][PYTHON] Python: connect client lost column data with pyarrow.Table.to_pylist

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Python: connect client should not use pyarrow.Table.to_pylist to transform fetched data.
For example:
the data in pyarrow.Table show below.

```
pyarrow.Table
key: string
order: int64
nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW): string
nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW): string
nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW): string
----
key: [["a","a","a","a","a","b","b"]]
order: [[0,1,2,3,4,1,2]]
nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW): [[null,"x","x","x","x",null,null]]
nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW): [[null,"x","x","x","x",null,null]]
nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW): [[null,null,"y","y","y",null,null]]
```

The table have five columns show above.
But the data after call pyarrow.Table.to_pylist() show below.

```
[{
	'key': 'a',
	'order': 0,
	'nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': None,
	'nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': None
}, {
	'key': 'a',
	'order': 1,
	'nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': 'x',
	'nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': None
}, {
	'key': 'a',
	'order': 2,
	'nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': 'x',
	'nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': 'y'
}, {
	'key': 'a',
	'order': 3,
	'nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': 'x',
	'nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': 'y'
}, {
	'key': 'a',
	'order': 4,
	'nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': 'x',
	'nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': 'y'
}, {
	'key': 'b',
	'order': 1,
	'nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': None,
	'nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': None
}, {
	'key': 'b',
	'order': 2,
	'nth_value(value, 2) OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': None,
	'nth_value(value, 2) ignore nulls OVER (PARTITION BY key ORDER BY order ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)': None
}]
```

There are only four columns left.

### Why are the changes needed?
Fix the bug that connect client lost column data with `pyarrow.Table.to_pylist`.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
Manual tests.

### CODE REVIEW

**Comment by zhengruifeng** [comment]
due to the duplicated column names?

can you update the example with a simpler one?

File: python/pyspark/sql/tests/connect/test_parity_functions.py [hunk 1]
```diff
@@ -90,8 +90,6 @@ def test_nested_higher_order_function(self):
     def test_np_scalar_input(self):
         super().test_np_scalar_input()
 
-    # TODO(SPARK-41904): Fix nth_value value
-    @unittest.skip("Fails in Spark Connect, should enable.")
     def test_nth_value(self):
```
Comment (zhengruifeng) [inline python/pyspark/sql/tests/connect/test_parity_functions.py:93]: is this PR related to https://github.com/apache/spark/pull/39456?

File: python/pyspark/sql/connect/dataframe.py [hunk 2]
```diff
@@ -1147,18 +1147,28 @@ def collect(self) -> List[Row]:
...
-            _dict: Dict[Any, Any] = {}
-            for k, v in row.items():
+        i = 0
+        while i < table.num_rows:
+            keys: List[Any] = []
+            values: List[Any] = []
+            j = 0
+            while j < table.num_columns:
+                k = table.column_names[j]
+                v = table.column(j).to_pylist()[i]
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/dataframe.py:1157]: is it possible that `table.column(j).to_pylist()` also have this issue?

`table.column(j)` could also be a struct type

File: python/pyspark/sql/connect/dataframe.py [hunk 3]
```diff
@@ -1147,18 +1147,28 @@ def collect(self) -> List[Row]:
         table = self._session.client.to_table(query)
 
         rows: List[Row] = []
-        for row in table.to_pylist():
-            _dict: Dict[Any, Any] = {}
-            for k, v in row.items():
+        i = 0
+        while i < table.num_rows:
+            keys: List[Any] = []
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/dataframe.py:1152]: keys are list of columns names,  can reuse it across different rows?

File: python/pyspark/sql/connect/dataframe.py (hunk 2 above)
Comment (beliefer) [reply python/pyspark/sql/connect/dataframe.py:1157]: table.column(j) is instance of `pyarrow.ChunkedArray`, but  table is instance of `pyarrow.Table`

**Comment by zhengruifeng** [comment]
would you mind also adding some tests in `test_connect_basic`?

File: python/pyspark/sql/connect/dataframe.py (hunk 2 above)
Comment (beliefer) [reply python/pyspark/sql/connect/dataframe.py:1157]:
```
>>> import pyarrow as pa
>>> n_legs = pa.chunked_array([[2, 2, 4], [4, None, 100]])
>>> n_legs.to_pylist()
[2, 2, 4, 4, None, 100]
```

pyarrow.ChunkedArray doesn't  have schema.

Comment (zhengruifeng) [reply python/pyspark/sql/connect/dataframe.py:1157]: `pyarrow.ChunkedArray` does not hit this issue?

Comment (zhengruifeng) [reply python/pyspark/sql/connect/dataframe.py:1157]: got it

File: python/pyspark/sql/tests/connect/test_parity_functions.py (hunk 1 above)
Comment (beliefer) [reply python/pyspark/sql/tests/connect/test_parity_functions.py:93]: Yes. The bug will duplicate the fields for output.

File: python/pyspark/sql/connect/dataframe.py (hunk 3 above)
Comment (beliefer) [reply python/pyspark/sql/connect/dataframe.py]: OK.

**Comment by beliefer** [comment]
> would you mind also adding some tests in `test_connect_basic`?

It doesn't matter!

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/tests/connect/test_parity_functions.py [hunk 4]
```diff
@@ -90,8 +90,6 @@ def test_nested_higher_order_function(self):
     def test_np_scalar_input(self):
         super().test_np_scalar_input()
 
-    # TODO(SPARK-41904): Fix nth_value value
```
Comment (zhengruifeng) [inline python/pyspark/sql/tests/connect/test_parity_functions.py:93]: is this `nth_value` issue related?

if so, please also remove `def test_nth_value(self)`

Comment (beliefer) [reply python/pyspark/sql/tests/connect/test_parity_functions.py:93]: OK

**Comment by zhengruifeng** [comment]
the failed test should be irrelevant, merged into master, thank you @beliefer

**Comment by beliefer** [comment]
@zhengruifeng Thank you!

**Pull request closed without merging**